### PR TITLE
Additions to the recompiler.

### PR DESCRIPTION
**Pull request by Mou-Ikkai**
Added all the Registers I could find.
Also added the opcodes: Movzx_r32_rm8, Nopd, Setne_rm8, Sub_64_imm8.

--- Commit Batch 2 additions ---
Opcodes: Add_r32_rm32, Add_rm32_r32, And_EAX_imm32, Cmp_rm8_imm8, Cmp_rm32_imm8, Jle_rel8_64, Jle_rel32_64, Js_rel8_64, Js_rel32_64, Lea_r32_m, Mov_rm32_imm32, Mov_rm64_imm32, Sete_rm8

### CODE REVIEW

**Comment by Mou-Ikkai** [comment]
I'd also like to say I tested all these operations, and none of them threw an error or gave an unexpected behavior.

File: src/kernel/src/process/module/recompiler.rs [hunk 1]
```diff
@@ -158,11 +158,14 @@ impl<'input> Recompiler<'input> {
                 Code::Mov_rm8_r8 => (self.transform_mov_rm8_r8(i), false),
                 Code::Mov_rm32_r32 => (self.transform_mov_rm32_r32(i), false),
                 Code::Mov_rm64_r64 => (self.transform_mov_rm64_r64(i), false),
-                Code::Nop_rm16 | Code::Nop_rm32 | Code::Nopw => (self.preserve(i), false),
+                Code::Movzx_r32_rm8 => (self.preserve(i), false),
```
Comment (ultimaweapon) [inline src/kernel/src/process/module/recompiler.rs:161]: We can't preserve this one because the source operand may use RIP-relative, which we do not supported yet.

File: src/kernel/src/process/module/recompiler.rs [hunk 2]
```diff
@@ -158,11 +158,14 @@ impl<'input> Recompiler<'input> {
...
+                Code::Nop_rm16 | Code::Nop_rm32 | Code::Nopd | Code::Nopw => (self.preserve(i), false),
                 Code::Pop_r64 => (self.preserve(i), false),
                 Code::Pushq_imm32 => (self.preserve(i), false),
                 Code::Push_r64 => (self.preserve(i), false),
                 Code::Retnq => (self.preserve(i), true),
+                Code::Setne_rm8 => (self.preserve(i), false),
```
Comment (ultimaweapon) [inline src/kernel/src/process/module/recompiler.rs:167]: Same here.

File: src/kernel/src/process/module/recompiler.rs [hunk 3]
```diff
@@ -647,8 +660,22 @@ impl<'input> Recompiler<'input> {
...
+            Register::R9 => Register::R8,
+            Register::R10 => Register::R11,
+            Register::R11 => Register::R10,
+            Register::R12 => Register::R13,
+            Register::R13 => Register::R12,
+            Register::R14 => Register::R15,
+            Register::R15 => Register::R14,
+            Register::RDI => Register::RSI,
+            Register::RSI => Register::RDI,
+            Register::RBP => Register::RSP,
```
Comment (ultimaweapon) [inline src/kernel/src/process/module/recompiler.rs:673]: We cannot use `RSP` as a general purpose register because it must be pointing to the stack all the time.

File: src/kernel/src/process/module/recompiler.rs [hunk 4]
```diff
@@ -647,8 +660,22 @@ impl<'input> Recompiler<'input> {
...
+            Register::RSP => Register::RBP,
```
Comment (ultimaweapon) [inline src/kernel/src/process/module/recompiler.rs:674]: `keep` argument should never be `RSP` because we should not have any cases to transform any instructions that writing to it.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]
You did a good job :) Just some issues.

**Comment by Mou-Ikkai** [comment]
I fixed these issues, however, I'm working on a part 2 for this, so hold until next commit.

**Comment by Mou-Ikkai** [comment]
@ultimaweapon Have fun! I would keep going but... I'm stuck on Imul_r32_rm32_imm8

File: src/kernel/src/process/module/recompiler.rs [hunk 5]
```diff
@@ -135,10 +135,15 @@ impl<'input> Recompiler<'input> {
 
             // Transform instruction.
             let (size, end) = match i.code() {
+                Code::Add_r32_rm32 => (self.transform_add_r32_rm32(i), false),
                 Code::Add_rm8_r8 => (self.transform_add_rm8_r8(i), false),
+                Code::Add_rm32_r32 => (self.transform_add_rm32_r32(i), false),
                 Code::Add_rm64_imm8 => (self.transform_add_rm64_imm8(i), false),
+                Code::And_EAX_imm32 => (self.transform_and_eax_imm32(i), false),
```
Comment (ultimaweapon) [inline src/kernel/src/process/module/recompiler.rs:142]: This one can be preserved because it does not access the memory.

File: src/kernel/src/process/module/recompiler.rs [hunk 6]
```diff
@@ -546,6 +722,46 @@ impl<'input> Recompiler<'input> {
         }
     }
 
+    fn transform_movzx_r32_rm8(&mut self, i: Instruction) -> usize {
+        // Check if first operand use RIP-relative.
+        if i.op0_kind() == OpKind::Memory && i.is_ip_rel_memory_operand() {
```
Comment (ultimaweapon) [inline src/kernel/src/process/module/recompiler.rs:727]: We need to check the second operand, not first.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]
Impressive! Just some small mistakes.

**Review by ultimaweapon — APPROVED** [review]
